src/Dictionaries/DirectDictionary.cpp

```diff
@@ -164,6 +164,138 @@ Columns DirectDictionary<dictionary_key_type>::getColumns(
+    {
+        if (!request.shouldFillResultColumnWithIndex(attribute_index))
+            continue;
+
+        auto & fetched_column_from_storage = fetched_columns_from_storage[attribute_index];
+        fetched_column_from_storage->reserve(requested_keys.size());
+    }
+
+    size_t fetched_key_index = 0;
+
+    Columns block_key_columns;
+    size_t dictionary_keys_size = dict_struct.getKeysNames().size();
+    block_key_columns.reserve(dictionary_keys_size);
+
+    QueryPipeline pipeline(getSourcePipe(key_columns, requested_keys));
+
+    PullingPipelineExecutor executor(pipeline);
+
+    Stopwatch watch;
+    Block block;
+    size_t block_num = 0;
+    size_t rows_num = 0;
+    while (executor.pull(block))
+    {
+        if (!block)
+            continue;
+
+        ++block_num;
+        rows_num += block.rows();
+        convertToFullIfSparse(block);
+
+        /// Split into keys columns and attribute columns
+        for (size_t i = 0; i < dictionary_keys_size; ++i)
+            block_key_columns.emplace_back(block.safeGetByPosition(i).column);
+
+        DictionaryKeysExtractor<dictionary_key_type> block_keys_extractor(
+            block_key_columns, arena_holder.getComplexKeyArena());
+        auto block_keys = block_keys_extractor.extractAllKeys();
+
+        for (size_t attribute_index = 0; attribute_index < request.attributesSize(); ++attribute_index)
+        {
+            if (!request.shouldFillResultColumnWithIndex(attribute_index))
+                continue;
+
+            const auto & block_column = block.safeGetByPosition(
+                dictionary_keys_size + attribute_index).column;
+            fetched_columns_from_storage[attribute_index]->insertRangeFrom(
+                *block_column, 0, block_keys.size());
+        }
+
+        for (size_t block_key_index = 0; block_key_index < block_keys.size(); ++block_key_index)
+        {
+            auto block_key = block_keys[block_key_index];
+            key_to_fetched_index[block_key] = fetched_key_index;
+            ++fetched_key_index;
+        }
+
+        block_key_columns.clear();
+    }
+
+    LOG_DEBUG(&Poco::Logger::get("DirectDictionary"),
+        "read {} blocks with {} rows from pipeline in {} ms",
+        block_num, rows_num, watch.elapsedMilliseconds());
+
+    Field value_to_insert;
+
+    size_t requested_keys_size = requested_keys.size();
+
+    auto result_columns = request.makeAttributesResultColumns();
+
+    size_t keys_found = 0;
+
+    for (size_t attribute_index = 0; attribute_index < result_columns.size(); ++attribute_index)
+    {
+        if (!request.shouldFillResultColumnWithIndex(attribute_index))
+            continue;
+
+        auto & result_column = result_columns[attribute_index];
+
+        const auto & fetched_column_from_storage = fetched_columns_from_storage[attribute_index];
+
+        result_column->reserve(requested_keys_size);
+        default_mask.resize(requested_keys_size);
```

Comment:
Here and in all other dictionaries I see that in case of multiple attributes we refill `default_mask` on each attribute with the same values. Consider filling `default_mask` only once on first attribute or just in a separate loop before iterating through attributes

Reply:
I simply just add a check whether mask filled before or not, then do an action accordingly. In a separate loop would be bad since searching via the key would be costly, and I am afraid some change would happen in ssd cache. Actually, I think the cost of filling the mask would be quite minor because filling the mask is filled sequentially or alternatively just add a check in the if condition which is not much better as adding more branches.